0-bootstrap/modules/groups/variables.tf

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+variable "create_groups_holder" {
+  description = "Contain the details of the Groups to be created"
+  type = object({
+    create_groups = bool
+    required_groups = object({
+      group_org_admins           = string
+      group_billing_admins       = string
+      billing_data_users         = string
+      audit_data_users           = string
+      monitoring_workspace_users = string
+    })
+    optional_groups = object({
+      gcp_platform_viewer      = string
+      gcp_security_reviewer    = string
+      gcp_network_viewer       = string
+      gcp_scc_admin            = string
+      gcp_global_secrets_admin = string
+      gcp_audit_viewer         = string
+    })
+  })
+  default = {
+    create_groups = false
+    required_groups = {
+      group_org_admins           = ""
+      group_billing_admins       = ""
+      billing_data_users         = ""
+      audit_data_users           = ""
+      monitoring_workspace_users = ""
+    }
+    optional_groups = {
+      gcp_platform_viewer      = ""
+      gcp_security_reviewer    = ""
+      gcp_network_viewer       = ""
+      gcp_scc_admin            = ""
+      gcp_global_secrets_admin = ""
+      gcp_audit_viewer         = ""
+    }
+  }
+
+  validation {
+    condition     = var.create_groups_holder.create_groups == true ? can(regex("[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?[.])+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?", var.create_groups_holder.required_groups.group_org_admins)) : true
```

Comment:
I am inclined to leave out regex email validation since its known this has to be a valid email - the provider/upstream can check for this.

Reply:
@bharathkkb sure it is known that it has to be a valid email, but we have been able to put specific error messages for each email that could be filled wrong before terraform start running any resource creation.

As we are using for_each to iterate through this data structure, it don't get clear where is the error for the user. But with a custom error_message, the user will have a more friendly way to understand visualize it since we have done a specific error_message for every specific key/value pair.

Reply:
@mauro-cit I would defer on adding this unless we have some signal. This validation would be better to add it at the provider layer where we can test the regex properly.